### ISSUE TITLE
Performance: Remove AMQP notifications from database layer

### ISSUE THREAD

**Issue by agger-magenta** [magenta-aps/mox#15]
A benchmark test (running on a slow virtual machine) created about 85,000 Bruger objects and 16,000 Organisation objects, app. 101,000 objects in all.

Completing this import with AMQP notifications enabled in the DB took 

212 minutes = 3h32min =~ 8 objects/sec

After commenting out the AMQP notifications and reinstalling the DB, this was reduced to

 38 minutes = 0h38min =~ 44 objects/sec

Conclusion: Sending notifications about object creation and updates from the database layer is a _very bad_ idea, performance-wise. 

In December 2016, we implemented the Log module which logs  all API access from the application layer (i.e., in the Flask API) using  AMQP messages. It would be very easy to a) use these messages directly for notifications, and/or b) send separate notifications only for write operations, if so desired.

To conclude, AMQP notifications should be removed from the database in the next release. There's an 82% performance gain for write operations.

**Comment by danchr** [comment]
We'll need to run this by @alexthirifays…

**Comment by JesperVadKristensen** [comment]
Technically speaking, isn’t the increase  in performance roughly speaking 500 percent? Or less dramatically “five times faster”?
Unless the 82 percent refers to something else you’re selling yourself short :) Anyway, just wanted to say congratulations for discovering this!

//Jesper Vad

Fra: Carsten Agger [mailto:notifications@github.com]
Sendt: 13. september 2017 15:39
Til: magenta-aps/mox <mox@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Emne: [magenta-aps/mox] Performance: Remove AMQP notifications from database layer (#15)


A benchmark test (running on a slow virtual machine) created about 85,000 Bruger objects and 16,000 Organisation objects, app. 101,000.

Completing this import with AMQP notifications enabled in the DB took

212 minutes = 3h32min =~ 8 objects/sec

After commenting out the AMQP notifications and reinstalling the DB, this was reduced to

38 minutes = 0h38min =~ 44 objects/sec

Concolusion: Sending notifications about object creation and updates from the database layer is a very bad idea, performance-wise.

In December 2016, we implemented the Log module which logs all API access from the application layer (i.e., in the Flask API) using AMQP messages. It would be very easy to a) use these messages directly for notifications, and/or b) send separate notifications only for write operations, if so desired.

To conclude, AMQP notifications should be removed from the database in the next release. There's an 82% performance gain for write operations.

**Comment by agger-magenta** [comment]
Well, it is a factor 5 in all cases.

Please note, though, that this benchmark was recorded with the log notifications also disabled.

If these are enabled, the import time is 50m =~ 33 objects/sec

From other measurements, we know that enabling logging in the API would cause the total import time to grow to 246 minutes, so it's still a performance boost of a factor 5.

Which means that while AMQP notifications from the database absolutely need to go, those from the application layer are negotiable but also represent a performance hit of about 31% - which is definitely and in every way better than 500%.

**Comment by danchr** [comment]
Just stumbled over “a RabbitMQ Exchange that publishes messages received from PostgreSQL Notifications.” I think it would allow us to use Postgres' built-in notification support, and relegate conversion to the AMQP process. Might be faster…

[gmr/pgsql-listen-exchange](https://github.com/gmr/pgsql-listen-exchange/)

**Comment by danchr** [comment]
One possible solution is to retain notifications, but simplify the API: we issue a notification, but instead of describing the change, we merely include the UUID affected. Any consumer will then need to query the REST API for the actual data, ensuring that our logging and access control guarantees remain in effect.

**Comment by agger-magenta** [comment]
Another possibility is that of doing it from the application layer/REST API. This is, according to the experiments described above, a but less costly.

**Comment by danchr** [comment]
I've been hacking away at my local LoRA install in order to get imports going as fast as possible. I was able write ~65k objects in just shy of 12 minutes, reaching 91.5 writes per second. I did this on my Mac under these circumstances:

- LoRA runner under gunicorn and PyPy with 8 workers and 8 threads.
- Client is PyPy with up to 10 concurrent connections to LoRA.
- PostgreSQL fsync() disabled.

Although the final item is very unsafe and unsuitable for production, it's nice to know how much we can squeeze out of LoRA. I'd consider this the best case for writing — I haven't benchmarked reading, which is arguably even more interesting.
